### PR TITLE
Generate using exported model and enable gemma2-2b in ExecuTorch

### PR DESCRIPTION
**Pull request by guangy10**
# What does this PR do?

Adding `generate` support for exported model.
Adding `gemma2-2b` to `ExecuTorch` with tests.
Adding an integration test for `gemma-2b` that we've enabled already. 

### Additional Test in `ExecuTorch`
Running `gemma2-2b` E2E:
```
cmake-out/examples/models/llama2/llama_main --tokenizer_path=tokenizer_gemma2.bin --model_path=gemma2.pte --prompt="My name is"
I 00:00:00.001356 executorch:cpuinfo_utils.cpp:62] Reading file /sys/devices/soc0/image_version
I 00:00:00.001425 executorch:cpuinfo_utils.cpp:78] Failed to open midr file /sys/devices/soc0/image_version
I 00:00:00.001431 executorch:cpuinfo_utils.cpp:158] Number of efficient cores 4
I 00:00:00.001434 executorch:main.cpp:65] Resetting threadpool with num threads = 6
I 00:00:00.005564 executorch:runner.cpp:65] Creating LLaMa runner: model_path=gemma2.pte, tokenizer_path=tokenizer_gemma2.bin
E 00:00:03.701808 executorch:tiktoken.cpp:79] invalid tiktoken line:
I 00:00:03.701845 executorch:runner.cpp:88] Failed to load tokenizer_gemma2.bin as a Tiktoken artifact, trying BPE tokenizer
I 00:00:03.767485 executorch:runner.cpp:94] Reading metadata from model
I 00:00:03.767512 executorch:runner.cpp:119] Metadata: get_vocab_size = 256000
I 00:00:03.767515 executorch:runner.cpp:119] Metadata: get_bos_id = 2
I 00:00:03.767517 executorch:runner.cpp:117] Methond use_sdpa_with_kv_cache not found, using the default value 0
I 00:00:03.767518 executorch:runner.cpp:119] Metadata: use_sdpa_with_kv_cache = 0
I 00:00:03.767520 executorch:runner.cpp:119] Metadata: get_n_eos = 1
I 00:00:03.767521 executorch:runner.cpp:117] Methond append_eos_to_prompt not found, using the default value 0
I 00:00:03.767522 executorch:runner.cpp:119] Metadata: append_eos_to_prompt = 0
I 00:00:03.767524 executorch:runner.cpp:119] Metadata: get_max_seq_len = 123
I 00:00:03.767525 executorch:runner.cpp:117] Methond enable_dynamic_shape not found, using the default value 0
I 00:00:03.767527 executorch:runner.cpp:119] Metadata: enable_dynamic_shape = 0
I 00:00:03.767529 executorch:runner.cpp:119] Metadata: use_kv_cache = 1
I 00:00:03.767575 executorch:runner.cpp:119] Metadata: get_n_bos = 1
I 00:00:03.767604 executorch:runner.cpp:167] RSS after loading model: 0.000000 MiB (0 if unsupported)
I 00:00:04.408489 executorch:runner.cpp:234] RSS after prompt prefill: 0.000000 MiB (0 if unsupported)
My name is Lale and I love to play with my dolls. I started to play with dolls at the age of two years old. My favorite activity is dancing. I would like to help people. I would like to travel to Spain. I like to be a vet. I love to help people. I would like to travel. I would like to be. I love to travel. I would like to be. I like to travel. I would like to be. I love to travel. I would like to be. I love to travel. I would like to be. I love to travel
I 00:00:23.188418 executorch:runner.cpp:246] RSS after finishing text generation: 0.000000 MiB (0 if unsupported)
PyTorchObserver {"prompt_tokens":4,"generated_tokens":118,"model_load_start_ms":1727310065089,"model_load_end_ms":1727310068851,"inference_start_ms":1727310068851,"inference_end_ms":1727310088272,"prompt_eval_end_ms":1727310069492,"first_token_ms":1727310069492,"aggregate_sampling_time_ms":180,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
I 00:00:23.188436 executorch:stats.h:84] 	Prompt Tokens: 4    Generated Tokens: 118
I 00:00:23.188438 executorch:stats.h:90] 	Model Load Time:		3.762000 (seconds)
I 00:00:23.188440 executorch:stats.h:100] 	Total inference time:		19.421000 (seconds)		 Rate: 	6.075897 (tokens/second)
I 00:00:23.188442 executorch:stats.h:108] 		Prompt evaluation:	0.641000 (seconds)		 Rate: 	6.240250 (tokens/second)
I 00:00:23.188444 executorch:stats.h:119] 		Generated 118 tokens:	18.780000 (seconds)		 Rate: 	6.283280 (tokens/second)
I 00:00:23.188446 executorch:stats.h:127] 	Time to first generated token:	0.641000 (seconds)
I 00:00:23.188480 executorch:stats.h:134] 	Sampling time over 122 tokens:	0.180000 (seconds)
```

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #33709
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker
@gante
@amyeroberts
@qubvel

### CODE REVIEW

**Comment by guangy10** [comment]
Why am I start getting `fetch_tests - Unauthorized` error on all the PRs I'm creating?

**Comment by LysandreJik** [comment]
This seems to be happening more often, @ydshieh would you know what might be happening? Otherwise I'll reach out to the CircleCI team, this is hindering work on the repo

**Comment by LysandreJik** [comment]
I've asked the CircleCI team @guangy10, very sorry for the inconvenience.

**Comment by ydshieh** [comment]
Might be related to 

> Allow CI could be run on private forked repositories (e.g. new model additions) (#33594)

But not happening on all external contributor's PRs. Strange

**Comment by ydshieh** [comment]
Could you first try ..?

https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions

**Comment by ydshieh** [comment]
Another thing to check

[If you're following the fork instead of the upstream repo](https://support.circleci.com/hc/en-us/articles/360008097173-Troubleshooting-why-pull-requests-are-not-triggering-jobs-on-my-organization)

```
A user who submits a pull request to your repository from a fork, but no pipeline is triggered with the pull request. This can happen when the user is following the project fork on their personal account rather than the project itself on CircleCI.

This will cause the jobs to trigger under the user's personal account. If the user is following a fork of the repository on CircleCI, we will only build on that fork and not the parent, so the parent’s PR will not get status updates. 

In these cases, the user unfollows their fork of the project on CircleCI. This will trigger their jobs to run under the organization when they submit pull requests. Those users can optionally follow the source project if they wish to see the pipelines.
```

**Comment by guangy10** [comment]
> Could you first try ..?
> 
> https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions

Weird. Used to work fine on my old PRs. Tried "Refresh Permission". Let's see if it can be unblocked

**Comment by guangy10** [comment]
> Another thing to check
> 
> [If you're following the fork instead of the upstream repo](https://support.circleci.com/hc/en-us/articles/360008097173-Troubleshooting-why-pull-requests-are-not-triggering-jobs-on-my-organization)
> 
> ```
> A user who submits a pull request to your repository from a fork, but no pipeline is triggered with the pull request. This can happen when the user is following the project fork on their personal account rather than the project itself on CircleCI.
> 
> This will cause the jobs to trigger under the user's personal account. If the user is following a fork of the repository on CircleCI, we will only build on that fork and not the parent, so the parent’s PR will not get status updates. 
> 
> In these cases, the user unfollows their fork of the project on CircleCI. This will trigger their jobs to run under the organization when they submit pull requests. Those users can optionally follow the source project if they wish to see the pipelines.
> ```

"In these cases, the user unfollows their fork of the project on CircleCI. " I have no idea how to unfollow my fork on CircleCI. I don't even see if I'm following transformers on CircleCI. I created a CircleCI account with exact email and linked to my github, and I can't see any project I'm following..

**Comment by guangy10** [comment]
https://support.circleci.com/hc/en-us/articles/360008097173-Troubleshooting-why-pull-requests-are-not-triggering-jobs-on-my-organization
> so please be sure "Build forked pull requests" is enabled in Project Settings > Advanced

@ydshieh Their Wiki is so bad. I can't find this setting from anywhere.

**Comment by ydshieh** [comment]
Hmm, weird CircleCI issue ...

Could you check a last time:

top-left (like in the image): organization : switch to your own org, follow `transformers` there.
Optionally: Then switch to `huggingface` (if you can find it) and follow `transformers` there

If still not working, I could try to push a commit.

Screenshot 2024-09-28 115104

**Comment by ydshieh** [comment]
I push a commit to trigger the CI jobs. It runs now. Got to say I have no idea what is wrong on the CircleCI side with the permission issue as other external contributors' PRs don't have it.

**Comment by guangy10** [comment]
Hi @ydshieh thank you so much for debugging this issue together with me. Per the message from CircleCI support team, it seems like the `TRANSFORMERS_CONTEXT` in the `.circleci/config.yml` is causing the permission issue, and I was suggested to remove it from the `config.yml`. It seems like it's newly added in #33594 last week. Still I'm not sure why it only affects my PRs but works for other users, but hope this message can give you more pointer to help resolve this issue on my fork and PRs. Thank you.


Screenshot 2024-09-30 at 11 42 35 AM

**Comment by guangy10** [comment]
> Hmm, weird CircleCI issue ...
> 
> Could you check a last time:
> 
> top-left (like in the image): organization : switch to your own org, follow `transformers` there. Optionally: Then switch to `huggingface` (if you can find it) and follow `transformers` there
> 
> If still not working, I could try to push a commit.
> 
> Screenshot 2024-09-28 115104

BTW, I tried this. From the top-left on that page, there is not any other org to switch to other than huggingface, and I got an "page not found" error when trying to click the "Projects" tab and check if I'm following `transformers` there.

  
Screenshot 2024-09-30 at 11 00 57 AM

**Comment by guangy10** [comment]
Since @ydshieh pushed a commit to trigger the CI and all CIs are green. Can I get a review on this PR? @amyeroberts @qubvel

File: src/transformers/integrations/executorch.py
Comment (ArthurZucker) [inline src/transformers/integrations/executorch.py:118]: Quick question on your vivison here, this would mean we have to maintain two versions of `generate`, do you think we could "avoid" that if you pass self.key_cache and self.value_cache to the original generate function? 🤗

File: tests/models/gemma2/test_modeling_gemma2.py
Comment (ArthurZucker) [inline tests/models/gemma2/test_modeling_gemma2.py:214]: to revert!

**Review by ArthurZucker — COMMENTED** [review]
Thanks! 🤗 asked a question for general direction as I am wondering if there is a way for us to improve our generate make it a bit more compatible !

**Comment by ydshieh** [comment]
Hi @guangy10 Thank you for the message and contacting CircleCI team on your side too. I also think it is from #33594 from the beginning of seeing this issue. But as you mentioned, only a few (external) contributors face this issue, and the answer from CircleCI attached above is somehow confusing (i.e. it doesn't explain what **actually** causes it).

I will create a new personal github account and see if I can reproduce and come up with a solution.

**Comment by ydshieh** [comment]
@guangy10 Could you go https://app.circleci.com/home/ and see what you get there? 🙏

**Comment by ydshieh** [comment]
Well I am able to reproduce now 

https://github.com/huggingface/transformers/pull/33850

**Comment by ydshieh** [comment]
Opened a PR #33866 for this CircleCI issue

File: src/transformers/integrations/executorch.py
Comment (guangy10) [reply src/transformers/integrations/executorch.py:118]: I think the goal of the `generate` here is slightly different. This simple generate function is created for validating the correctness of the exported program, so it's used only in tests to guard on the exported program for each enabled model. Typically users won't directly take the exported program into product for inference/generate as it's just an intermediate representation to unlock other possibilities, that is, users should either further lower the exported program with ExecuTorch for running on-device (better perf) or switch to use the compiled program for training or server-side inference.

Comment (guangy10) [reply src/transformers/integrations/executorch.py:118]: I think it's better to keep this PR clean to still focus on model enablement using this simple generate. I could try follow up in a separate PR to see if the original `generate` function can be reused. The signature of the exported program is different than the eager model, so I guess we may need to create an adaptor for it other than passing the cache object to the `generate`. Can you point to me which `generate` is the original one?

File: tests/models/gemma2/test_modeling_gemma2.py
Comment (guangy10) [reply tests/models/gemma2/test_modeling_gemma2.py:214]: Aha, forget to change this back. Good catch!

**Comment by guangy10** [comment]
> Opened a PR #33866 for this CircleCI issue

@ydshieh Thanks a lot. Really appreciate the quick fix! Let me rebase on top of the fix

**Comment by guangy10** [comment]
> * Add the `run-slow` label to the PR

@ydshieh I don't see there is a way for me to add `run-slow` label to the PR. Is there one?

**Comment by ydshieh** [comment]
Are you able to see what is shown below?

Screenshot 2024-10-01 200227

**Comment by guangy10** [comment]
> Are you able to see what is shown below?
> 
> Screenshot 2024-10-01 200227

No, that's why I'm asking if the message is asking me to do so or the repo maintainers. Also wondering if we have a something like `pytorchbot` that can do labeling for any user w/o having write permission to the repo.
 
Screenshot 2024-10-01 at 11 29 21 AM

**Comment by ydshieh** [comment]
Thanks for the feedback. We will improve the stuffs. For now I just added it manually here.

**Comment by guangy10** [comment]
@ArthurZucker @amyeroberts could you help review for this PR?

**Comment by ydshieh** [comment]
Don't forget to push an (empty) commit with message: `[run_slow] gemma, gemma2`, thank you

File: tests/models/gemma/test_modeling_gemma.py
Comment (ArthurZucker) [inline tests/models/gemma/test_modeling_gemma.py:894]: is it supported to infer with a different prompt / batch of prompt?

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗

**Comment by guangy10** [comment]
Comments addressed.
Does it require a 2nd reviewer in order to merge?

**Comment by ArthurZucker** [comment]
Nope, just waiting for the CIs right now!

File: tests/models/gemma/test_modeling_gemma.py
Comment (guangy10) [inline tests/models/gemma/test_modeling_gemma.py:848]: @ArthurZucker I see a test failure may be related to this. I copied this over from another model with additional texts added, feel free to truncate it to a shorter message and push a new commit

Comment (ArthurZucker) [reply tests/models/gemma/test_modeling_gemma.py:848]: Ah yeah won't have time to update them tonight! But yeah we can merge otherwise!

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py:848]: @ArthurZucker Let me know if there is anything I can do to merge this PR

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: I will check this afternoon and merge

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: gemma: all passing

gemma2: `test_export_static_cache`

I am getting the following (the output is the same as the prompt), and I am not sure that is expected.

```
(Pdb) ep_generated_text
['Hello I am doing']
(Pdb) ep_generated_ids
tensor([[   2, 4521,  590, 1144, 3900,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0]])

```

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: For gemma, it could generate 

```
(Pdb) ep_generated_text
['Hello I am doing a project on the 1990s and I need to know what the most popular music was in the 1990s. I have looked on the internet and I have found']

```

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: @ydshieh thank you for helping on merging this PR. Are you running the test on CPU or GPU? Actually the export test is expected to run on CPU only, however, the entire `Gemma2IntegrationTest` is decorated `@require_torch_gpu`. To verify the correctness locally, if you comment out `@require_torch_gpu`, then run `RUN_SLOW=1 pytest tests/models/gemma2/test_modeling_gemma2.py -k test_export_static_cache`, it shows
```
tests/models/gemma2/test_modeling_gemma2.py::Gemma2IntegrationTest::test_export_static_cache huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:01<00:00,  1.98it/s]
PASSED
======================================================== 1 passed, 388 deselected, 4 warnings in 25.10s ========================================================
```

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: @ydshieh Yes, happy to explain more. The exported program is an intermediate artifact that should not be directly used for product or anything. Instead users typically should take this intermediate artifact and further lower it to ExecuTorch which will covers the hardware delegates, op fusion, memory planning, etc. So the starting point (i.e. export) is always on CPU. This is different from the torch compiled artifacts. So the reason I'm adding a `generate` util in this PR is to validate the correctness of the exported program and ensure it can be further consumed in the downstream ExecuTorch stack. This is clearly documented in the util API itself. So for this test the perf doesn't matter, only correctness maters, so we don't expect to run this test on GPU. And it may produce weird output even if it may work in some cases.

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: I thought it may be okay to run the test on a machine configured with GPU because the `device=cpu` is already hard-coded in the test itself, maybe I'm wrong. I'm wondering if there is a tag to mark this test cpu-only?

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: Thank you @ydshieh ! Once this PR is merged I can start scaling the model coverage to other models utilizing the same util function added in this PR.

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: Could you share your torch version ..?

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: I finally get it pass (on our runner) by changing to

> attn_implementation = None

No idea why `sdpa` always get the original prompt as output.

Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: I will push a commit and merge tomorrow morning.

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: > Could you share your torch version ..?

Mine is a dev version `2.6.0.dev20241007`. What is the torch version where the test fails?

>  attn_implementation = None

It doesn't look correct. If we set it to `None`, what is the default attention being used? If I recall correctly, SPDA is the only attention impl that supports StaticCache.

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: Yeah, torch version seem to be the issue. Let me dig into it and will update here shortly

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: Okay, I can confirm that exporting `gemma2` model will require `torch==2.5.0` in order to work correctly.
I also verified that running the test on `torch==2.4.1` or `torch==2.0.0` will also get the original prompt as output.

Here are the detailed package info:
```
pip list | grep torch
executorch                         0.5.0a0+f8cec53
executorchcoreml                   0.0.1
torch                              2.5.0
torchaudio                         2.5.0
torchvision                        0.20.0
```

`RUN_SLOW=1 pytest tests/models/gemma2/test_modeling_gemma2.py -k test_export_static_cache -v`
```
=========================================================================================== test session starts ===========================================================================================
platform darwin -- Python 3.10.13, pytest-7.2.0, pluggy-1.0.0 -- /Users/guangyang/miniconda3/envs/executorch/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/guangyang/transformers/.hypothesis/examples')
rootdir: /Users/guangyang/transformers, configfile: pyproject.toml
plugins: cov-4.1.0, anyio-4.4.0, xdist-3.3.1, hypothesis-6.84.2
collected 389 items / 388 deselected / 1 selected

tests/models/gemma2/test_modeling_gemma2.py::Gemma2IntegrationTest::test_export_static_cache PASSED                                                                                                 [100%]
```

Comment (guangy10) [reply tests/models/gemma/test_modeling_gemma.py]: I updated the PR to bump up the required `torch` version to `2.5.0`. 

`torch==2.5.0` is not available publicly until Oct 17, 2024. We can either merge this PR as-is, the `test_export_static_cache` for `gemma2` will be skipped in the CI until `torch 2.5.0` is available in `transformers`. Or we can defer `gemma2` enablement until `torch 2.5.0` is available in `transformers` by splitting the `gemma2` related code out of this PR and merge the rest. Personally I would prefer former, but let me know how you want to proceed. cc: @ArthurZucker @ydshieh

**Comment by guangy10** [comment]
Debug the `gemma2` export issue and update the required `torch` version for it.

File: tests/models/gemma/test_modeling_gemma.py
Comment (ydshieh) [reply tests/models/gemma/test_modeling_gemma.py]: Great! We are still using python 3.8 and torch 2.4.1. I plan to switch to python 3.9 at the end of October. Before that, python 2.5 won't be available in the system. We will think about that.

We can merge this PR as it is. Thank you for checking (it works indeed).

> If we set it to None, what is the default attention being used?

Just the native implementation of attention (in the modeling files) using torch operators.

> If I recall correctly, SPDA is the only attention impl that supports StaticCache.

The native implementation of attention also works with `StaticCache` I believe.

**Pull request merged**
